core/src/main/scala/com/softwaremill/react/kafka_api_opt1/Consumer.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt1
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.NotUsed
+import akka.stream.Graph
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.Shape
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.common.TopicPartition
+
+object Consumer {
+
+  // FIXME those should probably be real case classes (or value classes)?
+  type CommitMsg = Map[TopicPartition, OffsetAndMetadata]
+  type CommitConfirmation = Future[CommitMsg]
+
+  case class ConsumerShape[K, V](
+      commit: Inlet[CommitMsg],
+      messages: Outlet[ConsumerRecord[K, V]],
+      confirmation: Outlet[CommitConfirmation]
```

Comment:
> I like how simple Option 3 (but I have to write one big message about stopping :)), but I worry about introducing new abstraction level on top of existing kafka client API without having access to low level stuff in terms of streams.

There I disagree. We have for example a file Sink but it is no way a replacement for generic file APIs. It gives a tool to do the most common file operation with streams: append. There is already a low level API for files, just as there is already a low level API for Kafka: the official client. The reason why one would want a pre-packaged Kafka Akka Streams adapter is to embed it in a larger streaming context, and as such composability is a primary concern. Option 1 composes very badly as it needs awkward boilerplate and weird streams setup.

Just to give a real life example flow inspired by real-world:
 - consume from Kafka queue
 - transform then enrich messages from DB. This can be one-to-many transformation, aka `mapConcat` (i.e. you need can only commit the original message if *all* derivative messages has been processed)
 - route the resulting messages to one or more of a set of backends (DB, JMS, other Kafka topic), ack original message once all these acked back.

Try to build this with mock implementations using Option 1 and Option 3.